### PR TITLE
CI: Build on Mac

### PR DESCRIPTION
**Pull request by ResearchDaniel**
Changes proposed in this PR:
 * Add mac to CI

### CODE REVIEW

File: .github/workflows/inviwo.yml [hunk 1]
```diff
@@ -23,14 +23,14 @@ jobs:
             targets: 'ALL_BUILD DOXY-Inviwo DOXY-Python package'
             installer: 'inviwo-installer-win'
             artifact: 'build/inviwo-v*.exe'
-         # - os: 'macos-latest'
-         #   triplet: 'x64-osx'
-         #   mono: 'mono'
-         #   cmake: '-DIVW_MODULE_HDF5=OFF'
-         #   generator: 'Ninja'
-         #   targets: 'ALL_BUILD package'
```
Comment (petersteneteg) [inline .github/workflows/inviwo.yml:31]: Why did you change the target name. It is called ALL_BUILD

File: .github/workflows/inviwo.yml [hunk 2]
```diff
@@ -23,14 +23,14 @@ jobs:
             targets: 'ALL_BUILD DOXY-Inviwo DOXY-Python package'
             installer: 'inviwo-installer-win'
             artifact: 'build/inviwo-v*.exe'
-         # - os: 'macos-latest'
-         #   triplet: 'x64-osx'
-         #   mono: 'mono'
-         #   cmake: '-DIVW_MODULE_HDF5=OFF'
-         #   generator: 'Ninja'
```
Comment (petersteneteg) [inline .github/workflows/inviwo.yml:30]: generator is not used anymore. I would remove generator and specify a preset instead.

File: CMakePresets.json [hunk 3]
```diff
@@ -67,18 +67,25 @@
     },
 
     {
-      "name": "msvc-dev",
```
Comment (petersteneteg) [inline CMakePresets.json]: I don't understand why you have renamed this?
We this preset adds is just the generator, then it should be name accordingly
adding x64 is pointless to use in the name since that is the only value we support. 
We can also remove this preset on other platforms with conditionals (https://cmake.org/cmake/help/latest/manual/cmake-presets.7.html#condition) then windows is also not needed. Hence MSVC should be in the name.

File: CMakePresets.json [hunk 4]
```diff
@@ -67,18 +67,25 @@
...
     {
-      "name": "msvc-dev-vcpkg",
+      "name": "x64-windows-dev-vcpkg",
       "displayName": "MSVC 2022 Using vcpkg",
       "description": "MSVC 2022 Development using vcpkg configuration",
-      "inherits" : ["vcpkg", "msvc-dev"]
+      "inherits" : ["vcpkg", "x64-windows-dev"]
+    },
+    {
+      "name": "x64-osx-dev-vcpkg",
```
Comment (petersteneteg) [inline CMakePresets.json]: no point in putting OSX in the name of this preset.

File: .github/workflows/inviwo.yml [hunk 5]
```diff
@@ -141,12 +149,14 @@ jobs:
         path: ${{ matrix.artifact }}
 
     - name: Upload doxygen
+      if: runner.os != 'Windows'
```
Comment (petersteneteg) [inline .github/workflows/inviwo.yml]: this should be runner.os == Windows. 
but I think it would be better to make this a setting in the build matrix...

File: .github/workflows/inviwo.yml [hunk 6]
```diff
@@ -116,16 +113,27 @@ jobs:
...
 
+    - name: Configure CMake
+      shell: bash
+      if: runner.os != 'Windows'
+      run: >
+        cmake -S inviwo -B build --preset='${{ matrix.triplet }}-dev-vcpkg'
+        -DVCPKG_TARGET_TRIPLET='${{ matrix.triplet }}'
+        -DCMAKE_BUILD_TYPE=${{ env.BUILD_TYPE }}
+        -DIVW_DOXYGEN_PROJECT=OFF
+        ${{ matrix.cmake }}
```
Comment (petersteneteg) [inline .github/workflows/inviwo.yml:125]: the Doxygen stuff can be moved into matrix.cmake then you don't have to repeat the step.

**Review by petersteneteg — CHANGES_REQUESTED** [review]
might enable ubuntu to...

File: .github/workflows/inviwo.yml (hunk 1 above)
Comment (ResearchDaniel) [reply .github/workflows/inviwo.yml:31]: No, it is not on Ninja

File: CMakePresets.json (hunk 3 above)
Comment (ResearchDaniel) [reply CMakePresets.json]: Ok, this was to avoid duplicating the cmake setup in the CI script.

File: .github/workflows/inviwo.yml (hunk 1 above)
Comment (ResearchDaniel) [reply .github/workflows/inviwo.yml:31]: https://cmake.org/cmake/help/git-stage/guide/user-interaction/index.html#id18

File: CMakePresets.json (hunk 3 above)
Comment (petersteneteg) [reply CMakePresets.json]: I use them extensively in the CMake GUI, so they should make clean sense there...

Comment (ResearchDaniel) [reply CMakePresets.json]: Changed it back and moved the cmake preset setup to the matrix cmake config instead.

File: .github/workflows/inviwo.yml (hunk 5 above)
Comment (ResearchDaniel) [reply .github/workflows/inviwo.yml]: I think that it would be better if we moved these steps to separate job(s)

File: .github/workflows/inviwo.yml (hunk 6 above)
Comment (ResearchDaniel) [reply .github/workflows/inviwo.yml:125]: Good point, fixed

File: CMakePresets.json (hunk 4 above)
Comment (ResearchDaniel) [reply CMakePresets.json]: fixed

**Comment by ResearchDaniel** [comment]
Kudos to https://github.com/keepassxreboot/keepassxc/commit/ab153a24ec2d5963192db63076bc0be1afa3ad95# which got me on the right track of running macdeployqt during packaging instead of after building and  https://github.com/openstreetmap/merkaartor/commit/a0404fded32e4e794d91b5c9e198ab8ac08b8c2b for how to setup the CPACK_PRE_BUILD_SCRIPTS do so.

File: .github/workflows/inviwo.yml [hunk 7]
```diff
@@ -13,39 +13,38 @@ jobs:
...
+        os: [windows-latest, macos-latest] #, ubuntu-20.04]
         include:
           - os: 'windows-latest'
             triplet: 'x64-windows'
             mono: ''
-            cmake: '-DIVW_MODULE_HDF5=ON'
-            generator: 'Visual Studio 17 2022'
+            cmake: '--preset msvc-dev-vcpkg -DIVW_MODULE_HDF5=ON -DIVW_DOXYGEN_PROJECT=ON 
+                    -DIVW_DOXYGEN_LATEX_PATH="C:/tools/TinyTeX/bin/win32/pdflatex.exe"'
+            cmake_preset: 'msvc-dev-vcpkg'
```
Comment (petersteneteg) [inline .github/workflows/inviwo.yml]: this was never used, since you also put in in the cmake variable...

File: apps/inviwo/packaging/packaging.cmake [hunk 8]
```diff
@@ -87,16 +87,29 @@ if(WIN32)
...
-        set(CPACK_GENERATOR           "TGZ;DragNDrop")
+        set(CPACK_GENERATOR           "DragNDrop")
         set(CPACK_DMG_DS_STORE        "${IVW_ROOT_DIR}/Resources/DS_mapp")
         set(CPACK_DMG_VOLUME_NAME     "${CPACK_PACKAGE_FILE_NAME}")
     else()
         set(CPACK_GENERATOR "TGZ")
     endif()
-
+    
+    get_target_property(_qmake_executable Qt${QT_VERSION_MAJOR}::qmake IMPORTED_LOCATION)
```
Comment (petersteneteg) [inline apps/inviwo/packaging/packaging.cmake]: why the odd _ prefix for the names?

File: cmake/deploy-osx.cmake.in [hunk 9]
```diff
@@ -0,0 +1,15 @@
+set(MACDEPLOYQT "@MACDEPLOYQT@")
```
Comment (petersteneteg) [inline cmake/deploy-osx.cmake.in:1]: Is this only for adding the qt libs/plugins or does this handle rpaths to? 
What about the rpath for other libs? does cmake fix that with the new settings? 
Did we not have some other code to install the qt libs, is that still active?

File: cmake/globalconfig.cmake [hunk 10]
```diff
@@ -230,6 +230,28 @@ endif()
...
 if(${CMAKE_SYSTEM_NAME} MATCHES "Darwin")
     add_definitions(-DDARWIN)
+
+    # Set RPATH (relative path) differently depending on build-time and install/package.
+    # At build time we use full paths to the generatated libraries.
+    # When packaging, however, we must use relative paths so that the app 
+    # can be used on other computers.
+    # See:
+    # https://gitlab.kitware.com/cmake/community/-/wikis/doc/cmake/RPATH-handling#always-full-rpath
+    set(CMAKE_BUILD_WITH_INSTALL_RPATH TRUE)
```
Comment (petersteneteg) [inline cmake/globalconfig.cmake:240]: does this replace the work that fixup_bundle did before? That did the rpath rewriting before..

**Review by petersteneteg — CHANGES_REQUESTED** [review]

File: .github/workflows/inviwo.yml (hunk 7 above)
Comment (ResearchDaniel) [reply .github/workflows/inviwo.yml]: Good catch

File: cmake/globalconfig.cmake (hunk 10 above)
Comment (ResearchDaniel) [reply cmake/globalconfig.cmake:240]: I think that fixup_bundle was only done on qt plugins before. But yes, it ensures that the RPATH for installed binaries are adjusted.

File: cmake/deploy-osx.cmake.in (hunk 9 above)
Comment (ResearchDaniel) [reply cmake/deploy-osx.cmake.in:1]: > Is this only for adding the qt libs/plugins or does this handle rpaths to?

It copies qt libs/plugins and adjusts the RPATHs to those for all the libraries
> What about the rpath for other libs? does cmake fix that with the new settings?

Yes, with the new settings the RPATH of all binaries are set to @loader_path and @loader_path/../Frameworks when packaged

> Did we not have some other code to install the qt libs, is that still active?

Yes, ivw_deploy_qt and ivw_qt_add_to_install 
**ivw_deploy_qt** is only active on Windows to be able to run inviwo outside of the dev environment. We might want to refactor this.

**ivw_qt_add_to_install** is probably not needed on mac/windows anymore. I have not tried to remove it. On linux we would need to use an open source tool (I think that it si called linuxdeployqt or something).

File: apps/inviwo/packaging/packaging.cmake (hunk 8 above)
Comment (ResearchDaniel) [reply apps/inviwo/packaging/packaging.cmake]: Fixed

File: cmake/deploy-osx.cmake.in (hunk 9 above)
Comment (petersteneteg) [reply cmake/deploy-osx.cmake.in:1]: https://github.com/inviwo/inviwo/blob/f37dcbdfda16646529283ae007339c3b95a447f8/cmake/installutils.cmake#L279

seems like that might do duplicated work then? should we use the same system on windows?

Comment (petersteneteg) [reply cmake/deploy-osx.cmake.in:1]: might as well use windeployqt on windows, and get rid of that crazy code to find the plugins in ivw_qt_add_to_install?

Comment (ResearchDaniel) [reply cmake/deploy-osx.cmake.in:1]: I am testing it on Mac now but don’t have a windows machine right now. Will have to hope that the CI build captures any errors :)

since we already run windeployqt after each build I think that it should be possible to remove all of the stuff, but otherwise it will probably work if we just run it again during packaging

Comment (ResearchDaniel) [reply cmake/deploy-osx.cmake.in:1]: Works on Windows now as well. You should checkout and test it :)
We might want to filter out which dlls to run windeployqt on, e.g., only inviwo-module dlls... However, it seems to work now so perhaps no need to mess with it more.

File: .github/workflows/inviwo.yml [hunk 11]
```diff
@@ -13,39 +13,39 @@ jobs:
   build:
     strategy:
       matrix:
-        os: [windows-latest] #, macos-latest, ubuntu-20.04]
+        os: [windows-latest, macos-latest] #, ubuntu-20.04]
```
Comment (petersteneteg) [inline .github/workflows/inviwo.yml]: is linux still failing? seemed like it was successful in the test?

File: .github/workflows/inviwo.yml [hunk 12]
```diff
@@ -89,20 +90,20 @@ jobs:
         -username "inviwo"
         -password "${{ secrets.GITHUB_TOKEN }}"
 
-    - name: Install Apt
+    - name: Apt Install, switch to gcc-11
       shell: bash
       if: runner.os == 'Linux'
       run: |
         sudo apt install libxinerama-dev libxcursor-dev xorg-dev libglu1-mesa-dev libglew-dev ninja-build doxygen-latex
```
Comment (petersteneteg) [inline .github/workflows/inviwo.yml]: could probably skip Doxygen here too

**Review by petersteneteg — DISMISSED** [review]
The windows installer seems to work for me

Did you disable linux again? seems from the CI run that is was active, but it is commented now?

File: .github/workflows/inviwo.yml (hunk 11 above)
Comment (ResearchDaniel) [reply .github/workflows/inviwo.yml]: It runs and works due to "include" below. Added this back

File: .github/workflows/inviwo.yml (hunk 12 above)
Comment (ResearchDaniel) [reply .github/workflows/inviwo.yml]: Fixed

**Comment by ResearchDaniel** [comment]
> Did you disable linux again? seems from the CI run that is was active, but it is commented now?

Active due to the include. Added the comment thingy now

**Review by petersteneteg — APPROVED** [review]
